TrueCraft.Client/MultiplayerClient.cs

```diff
@@ -62,58 +80,122 @@ public void RegisterPacketHandler(byte packetId, PacketHandler handler)
 
         public void Connect(IPEndPoint endPoint)
         {
-            Client.BeginConnect(endPoint.Address, endPoint.Port, ConnectionComplete, null);
+            SocketAsyncEventArgs args = new SocketAsyncEventArgs();
+            args.Completed += Connection_Completed;
+            args.RemoteEndPoint = endPoint;
+
+            if (!Client.Client.ConnectAsync(args))
+                Connection_Completed(this, args);
+        }
+
+        private void Connection_Completed(object sender, SocketAsyncEventArgs e)
+        {
+            if (e.SocketError == SocketError.Success)
+            {
+                Interlocked.CompareExchange(ref _connected, 1, 0);
+
+                Physics.AddEntity(this);
+
+                StartReceive();
+                QueuePacket(new HandshakePacket(User.Username));
+            }
+            else
+            {
+                throw new Exception("Could not connect to server!");
+            }
         }
 
         public void Disconnect()
         {
-            NetworkWorker.Abort();
-            new DisconnectPacket("Disconnecting").WritePacket(Stream);
-            Stream.BaseStream.Flush();
+            if (!Connected)
+                return;
+
+            Interlocked.CompareExchange(ref _connected, 0, 1);
+
+            QueuePacket(new DisconnectPacket("Disconnecting"));
+
+            Client.Client.Shutdown(SocketShutdown.Send);
```

Comment:
This is nonblocking. Make sure that the disconnect packet is sent before Disconnect returns.